contracts/utils/RolesBase.sol

```diff
@@ -281,12 +323,42 @@ contract RolesBase is IRolesBase {
         address toAccount,
         uint8[] memory roles
     ) internal virtual {
-        if (_getProposedRoles(fromAccount, toAccount) != _toAccountRoles(roles)) {
-            revert InvalidProposedRoles(fromAccount, toAccount, roles);
+        _acceptAccountRoles(fromAccount, toAccount, _toAccountRoles(roles));
+    }
+
+    /**
+     * @notice Internal function to accept roles transferred from another account.
+     * @dev Pending account needs to pass all the proposed roles.
+     * @dev Emits RolesRemoved and RolesAdded events.
+     * @param fromAccount The address of the current role
+     * @param accountRoles The account roles to accept
+     */
+    function _acceptAccountRoles(
+        address fromAccount,
+        address toAccount,
+        uint256 accountRoles
+    ) internal virtual {
+        if (_getProposedRoles(fromAccount, toAccount) != accountRoles) {
+            revert InvalidProposedRoles(fromAccount, toAccount, accountRoles);
         }
 
         _setProposedRoles(fromAccount, toAccount, 0);
```

Comment:
Shall we set proposed roles to 0 in transferRoles instead? Removes the edge case of having stale proposals if someone transfers directly

Reply:
i don't like the fact that transfers would now cost a bit more if they are not through a proposal (to set proposed roles from 0 to 0). There is also the edge case where I propose you become the OPERATOR and transfer the DISTRIBUTOR role to you after, causing the proposal of OPERATOR to be lost in that case, which would be a bug.